klient.c

```diff
@@ -0,0 +1,176 @@
+#include <sys/types.h>
+#include <sys/ipc.h>
+#include <sys/shm.h>
+#include <sys/sem.h>
+#include <signal.h>
+
+#define MY_MSG_SIZE 100 //Maksymalny rozmiar wiadomosci
+#define LOGIN_SIZE 20 //Maksymalny rozmiar loginu uzytkownika
+
+key_t shm_entries_key;
+key_t shm_state_key;
+key_t sem_entries_key;
+key_t sem_counter_key;
+int shm_entries_id;
+int shm_state_id;
+int sem_entries_id;
+int sem_counter_id;
+
+//Struktura dla wpisu
+struct entry{
+	char username[LOGIN_SIZE];
+	char message[MY_MSG_SIZE];
+	int likes;
+};
+
+//Struktura dla stanu serwera
+struct state{
+	int n; //Maksymalna liczba wpisow
+	int counter; //Aktualna liczba wpisow
+};
+
+struct entry *shm_entries;
+struct state *shm_state;
+
+int main(int argc, char* argv[]){
+	char bufor[MY_MSG_SIZE];
+	
+	//Tworzenie klucza i uzyskanie dostepu do segmentu pamieci wspoldzielonej (wpisy)
+	if((shm_entries_key= ftok(argv[1], 'R'))==-1){
+		perror("ftok()");
+		exit(1);
+	}
+	if((shm_entries_id= shmget(shm_entries_key, 0, 0))==-1){
+		perror("Error shmget()");
+		exit(1);
+	}
+
+	//Tworzenie klucza i uzyskanie dostepu do segmentu pamieci wspoldzielonej (stan serwera)
+	if((shm_state_key= ftok(argv[1], 'C'))==-1){
+		perror("ftok()");
+		exit(1);
+	}
+	if((shm_state_id= shmget(shm_state_key, 0, 0))==-1){
+		perror("Error shmget()");
+		exit(1);
+	}
+
+	//Dolaczenie pamieci wspolnej (wpisy)
+	shm_entries= (struct entry *) shmat(shm_entries_id, (void *)0, 0);
+	if(shm_entries==(struct entry*)-1){
+		perror("Error shmat()");
+		exit(1);		
+	}
+
+	//Dolaczenie pamieci wspolnej (stan serwera)
+	shm_state=(struct state*) shmat(shm_state_id, NULL, 0);
+	if(shm_state==(struct state*)-1){
+		perror("Error shmat()");
+		exit(1);		
+	}
+	
+	//Tworzenie klucza i uzyskanie dostepu do tablicy semaforow
+	sem_entries_key=ftok(argv[1], 'S');
+        if((sem_entries_id=semget(sem_entries_key, shm_state->n, 0666)) == -1) {
+                perror("Error semget()");
+                exit(1);
+        }
+
+	//Tworzenie klucza i uzyskanie dostepu do semafora licznika
+        sem_counter_key=ftok(argv[1], 'C');
+        if((sem_counter_id=semget(sem_counter_key, 1, 0666)) == -1) {
+                perror("Error semget()");
+                exit(1);
+        }
+
+
+	printf("Twitter 2.0 wita! (wersja C)\n");
+
+	struct sembuf sb;
+	struct sembuf sb_counter;
+
+	//Obsluga dodania nowego wpisu
+	if(argv[2][0]=='N'){
+		if(strlen(argv[3])>LOGIN_SIZE){
+			fprintf(stderr, "Login jest zbyt dlugi\n");
+			exit(1);
+		}
+
+		printf("[Wolnych %d wpisow(na %d)]\n", (shm_state->n)-(shm_state->counter), shm_state->n);
+		printf("Podaj swoj wpis: \n");
+		int count=read(0, bufor, MY_MSG_SIZE);
+		if(count>=MY_MSG_SIZE){
+			fprintf(stderr, "Wiadomosc jest zbyt dluga\n");
+			exit(1);
+		}
+
+		//Zablokowanie licznika wpisow
+		sb_counter.sem_num=0;
+		sb_counter.sem_op=-1;
+		sb_counter.sem_flg=0;
+		semop(sem_counter_id, &sb_counter, 1);
+		
+		//Aktualizacja licznika wpisow
+		int current_counter= shm_state->counter;
```

Comment:
Na tym etapie należy sprawdzić czy counter nie osiągnął już maksymalnej wartości. W takim przypadku należy wyświetlić wiadomość o braku możliwości dodania wpisu oraz zakończyć działanie programu

Reply:
Widzę, że ta obsługa już istnieje. Natomiast sprawdzenie występuje zbyt późno - to powoduje, że inkrementacja licznika nastąpi, nawet, gdy użytkownik otrzyma wiadomość o niemożliwości dodania kolejnego wpisu